### PR TITLE
Allow pure functions by passing context

### PR DESCRIPTION
**Pull request by gingerich**
As a composition utility, there are use cases beyond koa middleware. In some cases it's desirable to write pure middleware functions. This PR adds support for such pure functions.

### CODE REVIEW

**Comment by fl0w** [comment]
The issue I have is fragmented usage, context was previous bound to this then moved to passed context object and now it could optionally also be passed via next. I'm not convinced this would be a great addition for the Koa ecosystem.

**Comment by gingerich** [comment]
You're right that pure functions is not the use case of koa middleware. This change allows for broader applications than just koa middleware. Maybe for that reason it should be implemented separately, say in a fork of koa-compose.
However, I still think it's acceptable here because it doesn't change the way you write koa middleware functions. Downstream middleware does not care if context is passed implicitly with `next()` or explicitly with `next(ctx)` and upstream middleware should not expect a return value from next.

**Comment by jonathanong** [comment]
i don't see the benefit of this

File: index.js [hunk 1]
```diff
@@ -31,13 +31,14 @@ function compose (middleware) {
   return function (context, next) {
     // last called middleware #
     let index = -1
-    return dispatch(0)
-    function dispatch (i) {
+    return dispatch(0, context)
+    function dispatch (i, ctx = context) {
+      context = ctx
```
Comment (bertho-zero) [inline index.js:36]:
```suggestion
```

File: index.js [hunk 2]
```diff
@@ -31,13 +31,14 @@ function compose (middleware) {
...
       if (i <= index) return Promise.reject(new Error('next() called multiple times'))
       index = i
       let fn = middleware[i]
       if (i === middleware.length) fn = next
-      if (!fn) return Promise.resolve()
+      if (!fn) return Promise.resolve(ctx)
       try {
         return Promise.resolve(fn(context, dispatch.bind(null, i + 1)));
```
Comment (bertho-zero) [inline index.js:43]:
```suggestion
        return Promise.resolve(fn(context, dispatch.bind(null, i + 1, ctx)));
```

**Comment by jonathanong** [comment]
> there are use cases beyond koa middleware

this is out of scope of this repo :)

**Pull request closed without merging**